analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotTestThisWithIsOperator.cs

```diff
@@ -18,39 +18,44 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System;
 using System.Collections.Immutable;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
+using SonarAnalyzer.Extensions;
 using SonarAnalyzer.Helpers;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
     public sealed class DoNotTestThisWithIsOperator : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S3060";
+        private const string DiagnosticId = "S3060";
         private const string MessageFormat = "Offload the code that's conditional on this 'is' test to the appropriate subclass and remove the test.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(c =>
-                {
-                    var expression = (BinaryExpressionSyntax)c.Node;
-                    if (expression.Left is ThisExpressionSyntax)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, expression.GetLocation()));
-                    }
-                },
-                SyntaxKind.IsExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(c => Analyze(c, x => ((BinaryExpressionSyntax)x).Left), SyntaxKind.IsExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(c => Analyze(c, x => ((SwitchStatementSyntax)x).Expression), SyntaxKind.SwitchStatement);
+            context.RegisterSyntaxNodeActionInNonGenerated(c => Analyze(c, x => ((IsPatternExpressionSyntaxWrapper)x).Expression), SyntaxKindEx.IsPatternExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(c => Analyze(c, x => ((SwitchExpressionSyntaxWrapper)x).GoverningExpression), SyntaxKindEx.SwitchExpression);
+        }
+
+        private static void Analyze(SyntaxNodeAnalysisContext context, Func<SyntaxNode, ExpressionSyntax> retrieveExpression)
```

Comment:
Could be just `expression`. It's clear that it will return something when it's `Func`.
```suggestion
        private static void Analyze(SyntaxNodeAnalysisContext context, Func<SyntaxNode, ExpressionSyntax> expression)
```